### PR TITLE
refactor(controller): Netease Cloud Music API bundle

### PR DESCRIPTION
**Pull request by greenhat616**
## 目的
*  优化控制器加载流程
  *  [x] 自动注入子目录
  *  [x] 忽略 `_` 开头的目录/文件
* 重构网易云控制器树
  * 使用 Joi 进行参数校验
  * 拆分成多个子文件以便进行维护
  * 使用 Cache.remeber 方法简化控制流程

### CODE REVIEW

File: src/controllers/netease/redirect.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  if (result.status !== 200) {
+    ctx.status = Number.parseInt(result.status)
+    ctx.body = {
+      status: result.status,
+      message: '上游错误',
+      data: result.body,
+      ts: Date.now()
+    }
+    return
+  } else if (
```
Comment (AdaXH) [inline src/controllers/netease/redirect.js:33]: unnecessary else if,

File: src/controller.js [hunk 2]
```diff
@@ -13,7 +13,7 @@ function readDir (dir) {
     const stat = fs.statSync(filePath)
     if (stat.isDirectory()) {
       const subDirMaps = readDir(filePath)
-      if (subDirMaps && subDirMaps.length > 0) {
+      if (subDirMaps) {
```
Comment (AdaXH) [inline src/controller.js:16]: if (subDirMaps && subDirMaps.length)

File: src/controllers/netease/search.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    'PLAYLIST',
+    'SONG',
+    'USER',
+    'VIDEO',
+    'COMPLEX'
+  ).default('SONG'),
+  nocache: Joi.boolean().default(false)
+})
+
+async function getSearchDetail (params, ctx) {
```
Comment (AdaXH) [inline src/controllers/netease/search.js:28]: const { 
 keyword,
 limit,  
offset,
type,
} = params;

File: src/utils/crypto.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
+const crypto = require('crypto')
+
+exports.md5 = (rawData) => {
+  if (typeof rawData !== 'string') {
+    rawData = JSON.stringify(rawData)
```
Comment (AdaXH) [inline src/utils/crypto.js:5]: q:    rawData must be a JSON ?

File: src/controller.js (hunk 2 above)
Comment (greenhat616) [reply src/controller.js:16]: subDirMaps is a object

File: src/controllers/netease/search.js (hunk 3 above)
Comment (greenhat616) [reply src/controllers/netease/search.js:28]: ok, i will do it

File: src/utils/crypto.js (hunk 4 above)
Comment (greenhat616) [reply src/utils/crypto.js:5]: rawData might be series of types(number, json, etc)

File: src/controllers/netease/redirect.js (hunk 1 above)
Comment (greenhat616) [reply src/controllers/netease/redirect.js:33]: the result body like below:

{
cookies: xxx,
status: xxx, // HTTP Status code(performed by request)
body: {
code: xxx // the API ret code
}
}

Comment (greenhat616) [reply src/controllers/netease/redirect.js:33]: 🧐 any further suggestions?

**Comment by greenhat616** [comment]
还有最后一个 Summary 方法，明天起来搞=，=

File: src/cache.js [hunk 5]
```diff
@@ -99,7 +99,9 @@ class cache {
     let data = await this.get(key, toJSON)
     if (!data) { // data is empty
       data = await caller(...callerParams)
-      this.set(key, data, time) // async set
+      if (data) {
```
Comment (xiekeyi98) [inline src/cache.js:102]: If this judge is necessary? Maybe we already do this at line 100.

Comment (greenhat616) [reply src/cache.js:102]: The data was be reassigned by the func `caller`. `caller` is a func that can be filled as custom.
This is a usage :
```
Cache.remeber(key, time, callerFunc, [callerParams,[toJSON]])
Cache.rember(
  'test', // Redis Key
  60, // 60 seconds
  async (a, b, c) => { // a func was used to fetch data
      return a + b + c
   },
   ['a', 'b', 'c'],  // the array will be filled as the params of the caller func
   true // the result data(fetched by Redis Cache) will be parsed by `JSON.parse`
```

**Comment by greenhat616** [comment]
cc @hitokoto-osc/core-developers

**Review by AdaXH — APPROVED** [review]
LGTM

**Pull request merged**